cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1472,36 +1515,9 @@ class ObjectOutputStream final : public io::OutputStream {
       RETURN_NOT_OK(UploadPart("", 0));
     }
 
+    auto self = shared_from_this();
```

Comment:
Another problem is that we cannot call `shared_from_this` in dtor ( https://stackoverflow.com/questions/28338978/stdenable-shared-from-this-is-it-allowed-to-call-shared-from-this-in-destru ), so I separate a `FinishPartUploadAfterFlush`.